src/request.c

```diff
@@ -156,27 +156,113 @@ request_authorize_callback (GDBusInterfaceSkeleton *interface,
   return TRUE;
 }
 
+/* This is a bit ugly - we need to know where the options vardict is
+ * in the parameters for each request. Instead of inventing some
+ * complicated mechanism for each implementation to provide that
+ * information, just hardcode it here for now.
+ */
+static const char *
+get_token (GDBusMethodInvocation *invocation)
+{
+  const char *interface;
+  const char *method;
+  GVariant *parameters;
+  g_autoptr(GVariant) options = NULL;
+  const char *token = "t";
```

Comment:
> No, I actually mean "handles". That is what we call the out argument everywhere: @handle.

Ah, ok

> I've rewritten the "t" part to make it more clear what that was about: We want get_token() to always return a valid object path element, not NULL.

Ah, I see it better now, thanks for clarifying. However, unless I'm getting something fundamentally wrong (which considering my current state of mind is not unlikely) I don't think the new code is right either, looks to me like you'll be returning a pointer to potentially invalid memory. Let me explain it with a code excerpt, so that I'll make it even more embarrassing if I end up being wrong again:

```
static const char *
get_token (GDBusMethodInvocation *invocation)
{
  [...]
  g_autoptr(GVariant) options = NULL;
  const char *token;

  [...]

  if (strcmp (interface, "org.freedesktop.portal.Account") == 0)
    {
      options = g_variant_get_child_value (parameters, 1);
    }
  else if (strcmp (interface, "org.freedesktop.portal.Device") == 0)
    [...]

  if (options)
    g_variant_lookup (options, "token", "&s", &token);

  return token ? token : "t";
}

void
request_init_invocation (GDBusMethodInvocation *invocation, const char *app_id)
{
  [...]
  const char *token;

  request = g_object_new (request_get_type (), NULL);
  request->app_id = g_strdup (app_id);
  request->sender = g_strdup (g_dbus_method_invocation_get_sender (invocation));
  request->app_info = xdp_invocation_lookup_cached_app_info (invocation);

  token = get_token (invocation);
  sender = g_strdup (request->sender + 1);
  for (i = 0; sender[i]; i++)
    if (sender[i] == '.')
      sender[i] = '_';

  id = g_strdup_printf ("/org/freedesktop/portal/desktop/request/%s/%s", sender, token);

  [...]
}
```

If I'm getting this right, the `token = get_token (invocation)` line in `request_init_invocation()` is getting a `const char*` pointer to either the const string `"t"` in the .data section of the program (so that's ok) or a direct pointer to the string inside the `GVariant options` (see `g_variant_lookup (options, "token", "&s", &token)`) that lives inside the scope of `get_token()`, and gets freed after returning from that function.

So, my understanding is that `token` will be pointing to invalid (already freed) memory when you reference it from the `id = g_strdup_printf(...)` line, which doesn't seem right.

That's why I was proposing so use a `static` qualifier for the `token` variable in `get_token()`: to make sure that the pointer you returned pointed to valid memory. Now I just realized my suggestion was incomplete, since you'd also need to do something else like replace `"&s"` with a `"s"` in the format string when extracting the token, and at that point I wonder whether it wouldn't be simpler to just return a non-const char* in any case, and free it from `request_init_invocation()`, so that we don't mess with static pointers to strings which are probably not the best thing to use here in any case.

Does that make sense? If not, please feel free to point out if I said something silly in my reasoning, I might very well be overlooking something.